.github/workflows/docs.yml

```diff
@@ -0,0 +1,187 @@
+name: Docs
+
+on:
+  push:
+    branches:
+      - master
+    paths:
+      - "docs/**"
+      - "solidity/**"
+      - "!solidity/dashboard/**"
+  pull_request:
+  workflow_dispatch:
+
+jobs:
+  docs-detect-changes:
+    runs-on: ubuntu-latest
+    outputs:
+      path-filter: ${{ steps.filter.outputs.path-filter }}
+    steps:
+      - uses: actions/checkout@v2
+        if: github.event_name == 'pull_request' 
+      - uses: dorny/paths-filter@v2
+        if: github.event_name == 'pull_request' 
+        id: filter
+        with:
+          filters: |
+            path-filter:
+              - './((docs|(solidity/!dashboard))/**)'
+
+  docs-tex:
+    runs-on: ubuntu-latest
+    container: keepnetwork/texlive:3
+    needs: docs-detect-changes
+    if: |
+      github.event_name != 'pull_request'
+        || needs.docs-detect-changes.outputs.path-filter == 'true'
+    defaults:
+      run:
+        working-directory: ./docs
+    env:
+      DOCS_DIR: ${{ github.workspace }}/tmp/docs
+    steps:
+      - uses: actions/checkout@v2
+
+      # Step generates files based on Makefile config. Current config produces
+      # a number of .png files, which are then used to generate 
+      # relay-states.pdf. We want to publish only the resulting .pdf file.
+      - name: Generate TeX document files
+        run: |
+          make clean
+          make docs
+          mkdir -p $DOCS_DIR
+          mv relay-states.pdf $DOCS_DIR
+
+      # A push event is a master merge; deploy to primary bucket.
+      - if: github.event_name == 'push'
+        name: Upload generated files
+        uses: thesis/gcp-storage-bucket-action@v3.1.0
+        with:
+          service-key: ${{ secrets.DOCS_KEEP_NETWORK_UPLOADER_SERVICE_KEY_JSON }}
+          project: cfc-production
+          bucket-name: docs.keep.network
+          bucket-path: core
+          build-folder: tmp/docs
+
+      # A pull_request event is a PR; deploy to preview bucket.
+      - if: github.event_name == 'pull_request'
+        name: Upload preview of generated files
+        uses: thesis/gcp-storage-bucket-action@v3.1.0
+        with:
+          service-key: ${{ secrets.DOCS_KEEP_NETWORK_UPLOADER_SERVICE_KEY_JSON }}
+          project: cfc-production
+          bucket-name: docs.keep.network
+          bucket-path: core/${{ github.head_ref }}
+          build-folder: tmp/docs
+    
+  docs-solidity:
+    runs-on: ubuntu-latest
+    needs: docs-detect-changes
+    if: |
+      github.event_name != 'pull_request'
+        || needs.docs-detect-changes.outputs.path-filter == 'true'
+    env:
+      SOLIDITY_DOCS_DIR: ${{ github.workspace }}/docs/solidity
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Generate solidity docs
+        working-directory: ./solidity
+        run: |
+          npm ci
+          npx truffle compile
```

Comment:
Would definitely be interesting if we could reuse the build artifacts from the Solidity build here, fwiw. Maybe by attaching the Solidity docs build to the Solidity workflow rather than the docs workflow.

None of that is straightforward, so it's more of a “food for thought” thing---for now this is just fine.

Reply:
Will keep that in mind as something for consideration for the future, but merging the workflows might be hard to achieve, as the conditions on which we want to trigger Solidity and Docs are different (maybe this can be somehow accommodated, but likely with some code acrobatics, making the code harder to read and sustain).
But as I said, I will keep that in mind, maybe I'll think of some reasonable solution to this at later stage...